src/marqo/core/index_management/index_management.py

```diff
@@ -314,322 +197,121 @@ def get_all_indexes(self) -> List[MarqoIndex]:
+        the connection to zookeeper is lost, and more than one instance holds the lock. This should happen very rarely.
+        """
+        app_root_path = self.vespa_client.download_application(check_for_application_convergence=True)
+        app = VespaApplicationPackage(VespaApplicationFileStore(app_root_path))
 
-            tree.write(services_path)
-        else:
-            logger.warn(f"Schema {name} already exists in services.xml, nothing to add")
+        if check_configured and not app.is_configured:
+            raise ApplicationNotInitializedError()
 
-    def _remove_schema_from_services(self, app: str, name: str) -> None:
-        services_path = os.path.join(app, 'services.xml')
+        should_deploy = yield app
 
-        tree = ET.parse(services_path)
-        root = tree.getroot()
+        if should_deploy is None or should_deploy is True:
+            self.vespa_client.deploy_application(app_root_path)
+            self.vespa_client.wait_for_application_convergence()
 
-        documents_elements = root.findall("content/documents")
+        if should_deploy is not None:
+            yield
 
-        if len(documents_elements) > 1:
-            raise InternalError("Multiple documents elements found in services.xml. Marqo expects only one "
-                                "Vespa content cluster")
-        if len(documents_elements) == 0:
-            raise InternalError("No documents element found in services.xml")
+    @contextmanager
+    def _vespa_application_with_deployment_session(self, check_configured: bool = True):
+        """
+        A context manager that handles the deployment of a Vespa application package in a deployment session
+        This is a recommended way to deploy a Vespa application package. It leverages the optimistic locking mechanism
+        to avoid race conditions. Changes to the application package that do not touch any binary files should use
+        this context manager for deployment.
+        """
+        content_base_url, prepare_url = self.vespa_client.create_deployment_session()
+        store = ApplicationPackageDeploymentSessionStore(content_base_url, self.vespa_client)
+        app = VespaApplicationPackage(store)
 
-        documents_element = documents_elements[0]
+        if check_configured and not app.is_configured:
+            raise ApplicationNotInitializedError()
 
-        deleted = False
-        for document in documents_element.findall("document"):
-            if document.get("type") == name:
-                documents_element.remove(document)
+        should_deploy = yield app
 
-                if deleted:
-                    logger.warn(f"Multiple schemas with name {name} found in services.xml, removing all")
+        if should_deploy is None or should_deploy is True:
+            prepare_response = self.vespa_client.prepare(prepare_url)
+            # TODO handle prepare configChangeActions
```

Comment:
This `configChangeActions` field in the response from Vespa contains a list of actions (possibly empty) that must be performed in order to apply some config changes between the current active application and this next prepared application. hese actions are organized into three categories; restart, reindex, and refeed See https://docs.vespa.ai/en/reference/deploy-rest-api-v2.html#prepare-session for an example.  

What should we do when this list is not empty? 
- For requests modifying schemas (in the future), if we get any action in the list, should we reject the request?
- For bootstrap and rollback (in the future if they use deployment session after the Vespa bugfix), should we allow restart? 
@farshidz, what's your thought on this?